mlir/tools/mlir-rocm-runner/rocm-runtime-wrappers.cpp

```diff
@@ -353,122 +354,81 @@ extern "C" void mgpuMemCopy4DBF16(unsigned short *sourceAllocated, unsigned shor
-           int64_t filterStride2, int64_t filterStride3, float *inputAllocated,
-           float *inputAligned, int64_t inputOffset, int64_t inputSize0,
-           int64_t inputSize1, int64_t inputSize2, int64_t inputSize3,
-           int64_t inputStride0, int64_t inputStride1, int64_t inputStride2,
-           int64_t inputStride3, float *outputAllocated, float *outputAligned,
-           int64_t outputOffset, int64_t outputSize0, int64_t outputSize1,
-           int64_t outputSize2, int64_t outputSize3, int64_t outputStride0,
-           int64_t outputStride1, int64_t outputStride2, int64_t outputStride3,
-           int *layouts, int *layoutsAlligned, int64_t layoutOffset,
-           int64_t layoutSize, int64_t layoutStride, int32_t stride_h,
-           int32_t stride_w, int32_t padding_h, int32_t padding_w,
-           int32_t dilation_h, int32_t dilation_w) {
-
-  llvm::SmallVector<int64_t, 4> filterSizes(
-      {filterSize0, filterSize1, filterSize2, filterSize3});
-  llvm::SmallVector<int64_t, 4> filterStrides(
-      {filterStride0, filterStride1, filterStride2, filterStride3});
-  llvm::SmallVector<int64_t, 4> inputSizes(
-      {inputSize0, inputSize1, inputSize2, inputSize3});
-  llvm::SmallVector<int64_t, 4> inputStrides(
-      {inputStride0, inputStride1, inputStride2, inputStride3});
-  llvm::SmallVector<int64_t, 4> outputSizes(
-      {outputSize0, outputSize1, outputSize2, outputSize3});
-  llvm::SmallVector<int64_t, 4> outputStrides(
-      {outputStride0, outputStride1, outputStride2, outputStride3});
-
-  int batchSize, outChannelSize, inChannelSize;
-  int outHeightSize, outWidthSize;
-  int inHeightSize, inWidthSize;
-  int filterHeightSize, filterWidthSize;
-  int inBatchStride, inChannelStride, inHeightStride, inWidthStride;
-  int outBatchStride, outChannelStride, outHeightStride, outWidthStride;
-  int filterOutChanStride, filterInChanStride, filterHeightStride,
-      filterWidthStride;
-
-  // layouts[0..3]: filter layout
-  // layouts[4..7]: input layout
-  // layouts[8..11]: output layout
-  int *filterLayout = layouts;
-  int *inputLayout = layouts + 4;
-  int *outputLayout = layouts + 8;
+extern "C" void mcpuConv2d(int64_t rank1, void *f_ptr, int64_t rank2,
+                           void *i_ptr, int64_t rank3, void *o_ptr,
+                           int64_t rank4, void *f_layout, int64_t rank5,
+                           void *i_layout, int64_t rank6, void *o_layout,
+                           int32_t stride_h, int32_t stride_w,
+                           int32_t padding_h, int32_t padding_w,
+                           int32_t dilation_h, int32_t dilation_w) {
+
+  auto *filter = static_cast<StridedMemRefType<float, 4> *>(f_ptr);
+  auto *filterAllocated = filter->data + filter->offset;
+  auto filterSizes = llvm::ArrayRef<int64_t>(filter->sizes, rank1);
+  auto filterStrides = llvm::ArrayRef<int64_t>(filter->strides, rank1);
```

Comment:
It is quite good that you can come up with this way to work around unranked memref and convert it back to ranked. Due to the reason above, it is still not enough. 

What I asked for in the previous review comments shouldn't need this much maneuver, it should be as straightforward as three pointers to strided memref type: `StridedMemRefType<float, 4> * filterPtr, StridedMemRefType<float, 4> * inputPtr, StridedMemRefType<float, 4> * outputPtr`. With the argument type directly interpreted as target types, you can save off all those effort to reinterpret them in this routine.

After you finish this change, maybe file an issue to our repo and change all runtime wrappers not using `StridedMemRefType` to the usage of what I suggested.

Reply:
I think ranked memrefs would be interpreted into types used by the API in my previous version. StridedMemRefType needs unranked memref and type conversion.

Reply:
> StridedMemRefType needs unranked memref and type conversion.

Why is that? What's the problem of using ranked memref?

Reply:
A ranked memref is automatically interpreted as those explicit types that I used in the previous version. Below shows what is passed into mcpuConv2d if I use ranked memrefs and StridedMemRefType:

0x00007fffdb32f5fd in mcpuConv2d (filter=0x555555974420, input=0x555555974420, output=0x0, layout1=0x3, layout2=0x3, layout3=0x8, stride_h=128, stride_w=3072, padding_h=1024, padding_w=128, dilation_h=1, dilation_w=-641220592)
   
filter and input point to filter, output is actually filter offset, layout1-3 and stride_h are filter sizes, followed by filter strides, etc.

Reply:
We discussed offline. And prototyped using below APIs
- _mlir_ciface_print_memref_2d_f32
- print_memref_f32

The summary seems to be that unranked interface is favored over ranked one as it expose less implementation details. Should there be further test cases to work on, we should continue using unranked interface but mimic what has been done in `RunnerUtils.cpp` and make base implementation templatized.